src/netbase.cpp

```diff
@@ -48,36 +48,50 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
     ai_hint.ai_protocol = IPPROTO_TCP;
     // We don't care which address family (IPv4 or IPv6) is returned
     ai_hint.ai_family = AF_UNSPEC;
+
     // If we allow lookups of hostnames, use the AI_ADDRCONFIG flag to only
     // return addresses whose family we have an address configured for.
     //
     // If we don't allow lookups, then use the AI_NUMERICHOST flag for
     // getaddrinfo to only decode numerical network addresses and suppress
     // hostname lookups.
-    ai_hint.ai_flags = allow_lookup ? AI_ADDRCONFIG : AI_NUMERICHOST;
+    const int ai_flags = allow_lookup ? AI_ADDRCONFIG : AI_NUMERICHOST;
 
-    addrinfo* ai_res{nullptr};
-    const int n_err{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
-    if (n_err != 0) {
-        return {};
-    }
-
-    // Traverse the linked list starting with ai_trav.
-    addrinfo* ai_trav{ai_res};
     std::vector<CNetAddr> resolved_addresses;
-    while (ai_trav != nullptr) {
-        if (ai_trav->ai_family == AF_INET) {
-            assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in));
-            resolved_addresses.emplace_back(reinterpret_cast<sockaddr_in*>(ai_trav->ai_addr)->sin_addr);
+
+    auto callgetaddrinfo = [&](int flags) -> int {
```

Comment:
Unless I'm overlooking something, might be able to simplify/shorten the change by calling `getaddrinfo()` again when `n_err` is non-zero.  Would avoid the lambda function and might also get rid of the standalone `ai_flags`.

Something like:

```c++
addrinfo* ai_res{nullptr};
const int n_err{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
if (n_err != 0 && (ai_hint.ai_flags & AI_ADDRCONFIG) == AI_ADDRCONFIG) {
    // AI_ADDRCONFIG on some systems may exclude loopback-only addresses
    // If first lookup failed we perform a second lookup without AI_ADDRCONFIG
    ai_hint.ai_flags = (ai_hint.ai_flags & ~AI_ADDRCONFIG);
    const int n_err_retry{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
    if (n_err_retry != 0) {
        return {};
    }
}
```

Reply:
If pursuing this approach do we think it would be prudent to reset `ai_res` to a nullptr before the second call?

Reply:
Good thought.  At least from an initial look, what makes sense to me would be checking if it's already null and if not call `freeaddrinfo()` (`getaddrinfo()` handles allocation and `freeaddrinfo()` handles cleanup).  I suspect this could be somewhat overkill (or not happen if unsuccessful `getaddrinfo()` never sets ai_res to non-null), but we don't want a memory leak.  I'm still looking for guarantees in documentation (and we'd need to potentially compare guarantees for different platforms).

As a side note, `freeaddrinfo(NULL)` doesn't crash/segfault (at least on Debian/glibc), so the null check before `freeaddrinfo()` might be redundant (but defensive).

Reply:
I think it's overkill. In all the examples and documentation I've read, `freeaddrinfo()` is only called when `getaddrinfo()` succeeds.

Reply:
Probably, but the examples I've seen are all typically small and exit the program if `getaddrinfo()` returns non-zero.